### PR TITLE
feat(core): implement some ws send events

### PR DESCRIPTION
**Pull request by suneettipirneni**
**Please describe the changes this PR makes and why it should be merged:**

This implements guild member requests, voice state updating and presence updating for /core.

This PR is breaking as it converts `createClient` to a class.

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)
- This PR includes breaking changes (methods removed or renamed, parameters moved or removed)

### CODE REVIEW

File: packages/core/src/client.ts [hunk 1]
```diff
@@ -158,22 +164,101 @@ export interface ClientOptions {
...
+	public async requestGuildMembers(shardId: number, options: GatewayRequestGuildMembersData) {
+		const promise = new Promise<APIGuildMember[]>((resolve, reject) => {
+			const guildMembers: APIGuildMember[] = [];
+
+			const timeout = setTimeout(() => {
+				reject(new Error('Request timed out'));
+			}, 10_000);
+
+			const handler = ({ shardId: eventShardId, data }: MappedEvents[GatewayDispatchEvents.GuildMembersChunk][0]) => {
+				if (eventShardId !== shardId && data.guild_id !== options.guild_id) {
```
Comment (kyranet) [inline packages/core/src/client.ts:202]:
```suggestion
				if (eventShardId !== shardId || data.guild_id !== options.guild_id) {
```

Comment (kyranet) [reply packages/core/src/client.ts:202]: Also, shouldn't we check for nonce as well? Is the shardId check even relevant, considering guilds have only 1 shard?

**Review by kyranet — CHANGES_REQUESTED** [review]

**Review by almeidx — COMMENTED** [review]
The readme example needs to be updated to use the new Client class

File: packages/core/README.md [hunk 2]
```diff
@@ -59,7 +59,7 @@ client.on('interactionCreate', async ({ interaction, api }) => {
 });
 
 // Listen for the ready event
-client.on('ready', () => console.log('Ready!'));
+client.once('ready', () => console.log('Ready!'));
```
Comment (Syjalo) [inline packages/core/README.md:62]:
```suggestion
client.once(GatewayDispatchEvents.Ready, () => console.log('Ready!'));
```
An invalid example.
For L53
```ts
client.on(GatewayDispatchEvents.InteractionCreate, async ({ data: interaction, api }) => {
```

**Review by Syjalo — CHANGES_REQUESTED** [review]

File: packages/core/README.md (hunk 2 above)
Comment (Syjalo) [reply packages/core/README.md]: @suneettipirneni, there's also for line 53. I can't add suggestion for it.

**Review by kyranet — APPROVED** [review]

File: packages/core/src/client.ts [hunk 3]
```diff
@@ -158,22 +165,105 @@ export interface ClientOptions {
...
+
+	/**
+	 * Requests guild members from the gateway.
+	 *
+	 * @see {@link https://discord.com/developers/docs/topics/gateway-events#request-guild-members}
+	 * @param shardId - The id of the shard to request guild members from
+	 * @param options - The options for the request
+	 * @param timeout - The timeout for waiting for each guild members chunk event
+	 */
+	public async requestGuildMembers(shardId: number, options: GatewayRequestGuildMembersData, timeout = 10_000) {
```
Comment (vladfrangu) [inline packages/core/src/client.ts:195]: This should be redone to take in a guild id, not a shard id

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: packages/core/src/client.ts (hunk 3 above)
Comment (suneettipirneni) [reply packages/core/src/client.ts:195]: I removed the shardId parameter and made it take just the options data object.

File: packages/core/README.md [hunk 4]
```diff
@@ -46,11 +46,11 @@ const ws = new WebSocketManager({
 });
 
 // Create a client to emit relevant events.
-const client = createClient({ rest, ws });
+const client = new Client({ rest, ws });
 
 // Listen for interactions
 // Each event contains an `api` prop along with the event data that allows you to interface with the Discord REST API
-client.on('interactionCreate', async ({ interaction, api }) => {
+client.on(GatewayDispatchEvents.InteractionCreate, async ({ interaction, api }) => {
```
Comment (Syjalo) [inline packages/core/README.md:53]:
```suggestion
client.on(GatewayDispatchEvents.InteractionCreate, async ({ data: interaction, api }) => {
```
`interaction` is not a thing

**Review by Syjalo — CHANGES_REQUESTED** [review]

**Review by kyranet — COMMENTED** [review]
By the way, I was thinking that maybe we should probably support a version of `requestGuildMembers` which returns an async iterable, so we can do operations on members while they come thru. Obviously, `requestGuildMembers` would consume this async iterator and flatten the chunks.

**Review by iCrawl — APPROVED** [review]

File: packages/core/src/util/ws.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import type { Snowflake } from 'discord-api-types/v10';
+
+export function calculateShardId(guildId: Snowflake, shardCount: number) {
```
Comment (didinele) [inline packages/core/src/util/ws.ts:3]: I feel like this belongs in `@discordjs/util` 👀

**Review by didinele — CHANGES_REQUESTED** [review]

File: packages/core/src/util/ws.ts (hunk 5 above)
Comment (vladfrangu) [reply packages/core/src/util/ws.ts:3]: Agreed

**Review by vladfrangu — APPROVED** [review]

**Pull request merged**